### PR TITLE
circuits: zk-gadgets, zk-circuits: Refactor implementation after type refactor

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR makes the necessary changes to the `zk-gadgets` and `zk-circuits` crates to use the new `circuit_type` macro to derive types.

### Testing
- Unit tests pass, though CI may not

### CODE REVIEW

**Review by akirillo — APPROVED** [review]
confirming my understanding about a couple things + suggest a couple deletions, but overall looks correct & mergeable

File: circuits/src/zk_gadgets/elgamal.rs
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_gadgets/elgamal.rs:1]: We will need it again very shortly for the protocol fees + internal transfers, so I'll leave it in for now

File: circuits/src/zk_circuits/valid_match_mpc.rs [hunk 1]
```diff
@@ -75,65 +65,49 @@ pub struct ValidMatchMpcCircuit<'a, N: MpcNetwork + Send, S: SharedValueSource<S
...
+        cs.constrain(witness.order2.base_mint - witness.match_res.base_mint.clone());
 
         // Check that the direction of the match is the same as the first party's direction
-        cs.constrain(&matches.direction - &order1.side);
+        cs.constrain(witness.match_res.direction.clone() - witness.order1.side.clone());
 
         // Check that the orders are on opposite sides of the market. It is assumed that order
         // sides are already constrained to be binary when they are submitted. More broadly it
         // is assumed that orders are well formed, checking this amounts to checking their inclusion
         // in the state tree, which is done in `input_consistency_check`
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_circuits/valid_match_mpc.rs:93]: Yeah, this was an artifact that was never removed when we added proof-linking via commitments. Previously we hashed inside the ZK-MPC circuit and constrained this hash to equal a publicly known value consistent across proofs.

File: circuits/src/zk_gadgets/comparators.rs [hunk 2]
```diff
@@ -184,12 +175,12 @@ impl EqVecGadget {
 /// Returns a boolean representing a != b where 1 is true and 0 is false
 #[derive(Debug)]
 pub struct NotEqualGadget {}
-
 impl NotEqualGadget {
     /// Computes a != b
-    pub fn not_equal<L, CS>(a: L, b: L, cs: &mut CS) -> LinearCombination
+    pub fn not_equal<L1, L2, CS>(a: L1, b: L2, cs: &mut CS) -> LinearCombination
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_gadgets/comparators.rs:180]: Yeah I agree, I'll make this change for the equality gadgets, thanks!

File: circuits/src/zk_gadgets/fixed_point.rs [hunk 3]
```diff
@@ -80,9 +96,12 @@ impl FixedPoint {
...
     /// Commit to the fixed point variable as a public input in a given constraint system
-    pub fn commit_public<CS: RandomizableConstraintSystem>(&self, cs: &mut CS) -> FixedPointVar {
+    pub fn commit_public<CS: RandomizableConstraintSystem>(
+        &self,
+        cs: &mut CS,
+    ) -> FixedPointVar<Variable> {
         let repr = cs.commit_public(self.repr);
-        FixedPointVar { repr: repr.into() }
+        FixedPointVar { repr }
     }
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_gadgets/fixed_point.rs:105]: Yes, good catch, old artifact that I mindlessly type-refactored without thinking

File: circuits/src/zk_gadgets/merkle.rs [hunk 4]
```diff
@@ -467,14 +243,23 @@ pub(crate) mod merkle_test {
...
+    ) -> [Scalar; HEIGHT]
+    where
+        [(); HEIGHT]: Sized,
+    {
+        let mut res = Vec::with_capacity(HEIGHT);
+        for _ in 0..HEIGHT {
+            res.push(match leaf_index % 2 {
+                val @ 0..=1 => Scalar::from(val as u64),
+                _ => unreachable!("impossible evaluation mod 2"),
+            });
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_gadgets/merkle.rs:257]: done

File: circuits/src/zk_gadgets/select.rs [hunk 5]
```diff
@@ -359,29 +169,25 @@ mod cond_select_test {
...
+        let a_var = a.commit_public(&mut prover);
+        let b_var = b.commit_public(&mut prover);
+
+        // Selector = 1
+        let selector = Scalar::one().commit_public(&mut prover);
+        let res = CondSelectGadget::select(a_var, b_var, selector, &mut prover);
+
+        assert_eq!(a, prover.eval(&res));
 
-        // Invalid proof
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuits/src/zk_gadgets/select.rs:376]: I think the difference is that gadgets "compute" something whereas circuits combine these computations into "constraints" (as far as the abstraction goes). So the constraint satisfaction tests get subsumed in application level circuits. 

Practically speaking, we're only calling these selectors with outputs of other gadgets that are binary.

**Pull request merged**